request.lisp

```diff
@@ -658,20 +659,28 @@ Any encodings in Transfer-Encoding, such as chunking, are always performed."
                       (puri:uri-query uri) nil))
               (write-http-line "~A ~A ~A"
                                (string-upcase method)
-                               (if (and preserve-uri
-                                        (stringp unparsed-uri))
-                                   (trivial-uri-path unparsed-uri)
-                                   (puri:render-uri (if (and proxy
-                                                             (null stream)
-                                                             (not proxying-https-p)
-                                                             (not real-host))
-                                                        uri
-                                                        (make-instance 'puri:uri
-                                                                       :path (puri:uri-path uri)
-                                                                       :parsed-path (puri:uri-parsed-path uri)
-                                                                       :query (puri:uri-query uri)
-                                                                       :escaped t))
-                                                    nil))
+                               (let ((uri-string (if (and preserve-uri
+                                                          (stringp unparsed-uri))
+                                                     (trivial-uri-path unparsed-uri)
+                                                     (puri:render-uri (if (and proxy
+                                                                               (null stream)
+                                                                               (not proxying-https-p)
+                                                                               (not real-host))
+                                                                          uri
+                                                                          (make-instance 'puri:uri
+                                                                             :path (puri:uri-path uri)
+                                                                             :parsed-path (puri:uri-parsed-path uri)
+                                                                             :query (puri:uri-query uri)
+                                                                             :escaped t))
+                                                                      nil))))
+                                 (if encode-unicode-path-p
+                                     (with-output-to-string (*standard-output*)
+                                       (loop for c across uri-string
+                                             if (> (char-code c) 255)
+                                               ;; It's not a latin-1 character, so we need to encode it.
```

Comment:
URLs must only contain US-ASCII characters, everything else must be encoded.

Reply:
Drakma raised an exception when encountering URLs with Unicode characters in the path or the query parameters. To make URLs more accessible for non-English users, many websites have tried to incorporate Unicode characters in these sections of the URLs, even the HTTP protocol says a URL only contain US-ASCII characters.

I wonder whether we need to support it inside Drakma, if not, I'll try to revert related code change.

Reply:
Hello @jingtaozf,

I understand what you're trying to accomplish.  What I mean to say is that in the encoded URL, only US-ASCII characters are permitted, but you're checking for `(> (char-code c) 255)`, which would pass non-US-ASCII characters  as well.  There also is the issue of determining the correct encoding for those characters.  Nowadays, UTF-8 can mostly be assumed, but some web servers may actually try to use the Content-Type to determine the encoding.  Some experimentation will be needed, I think.

In any case, I'd recommend that you check for `(> (char-code c) 126)` and encode using percent encoding using UTF-8.

-Hans